File: frontend/maquette/harness/focus.py
"""R81 — what an assistive technology is told, and an audit cannot see.

Three things live here, and they share one property: `a11y.py` cannot see any
of them. Focus MOVING, a surface saying it is BUSY, and an error ANNOUNCING
itself are facts about a sequence, or about a state the audit is not driving.

WHY A RULE AND NOT THE ACCESSIBILITY AUDIT. `a11y.py` reads the markup of ONE
MOMENT. Focus management is a SEQUENCE: something was focused, a layer opened,
focus moved inside it, the rest of the frame stopped being reachable, the layer
closed, and focus went back where it came from. No static audit can observe any
of that, so the two instruments measure different things and neither replaces
the other.

WHAT FAILS WHEN THIS IS ABSENT, and each half fails on its own:

  · a layer that opens without taking focus leaves a keyboard on the page
    behind it — Tab walks a list the reader cannot see, and a screen reader
    keeps announcing a surface the interface has covered;
  · a layer that closes without GIVING focus back drops the caret at the top of
    the document, and the reader starts again.

THE BACKGROUND IS CHECKED BY `inert`, NOT BY `aria-hidden`. `aria-hidden` hides
a subtree from a screen reader and leaves every control in it tabbable, which is
worse than either half alone: the reader tabs into something no longer
announced.
"""
import asyncio
import sys

from common import BAR, Journal, open_page
from playwright.async_api import async_playwright

# One entry per layer this rule drives: the control that opens it, the layer's
# root, and a short name for the report.
LAYERS = (
    ("drawer", '[data-drawer="1"]', "#drawer"),
    ("sheet", '[data-sheet="utilisateur"]', "#sheet"),
    # A MIGRATED SCREEN, and it is here because leaving it out is exactly
    # how the manager shipped blind to five of its six layers: there is one
    # legacy screen carrying `id="screen"` and five React ones carrying no
    # id at all, so a rule that only drove the drawer and the sheet was
    # green over a manager selecting `#screen`.
    ("screen", '[data-mediasheet]', '[data-part="screen"][data-open]'),
)

# Where the tab order must stop while a layer is up. `#port` is the main region
# behind every layer; if it is still reachable, so is the whole page.
BACKGROUND = "#port"

FOCUS_STATE = """([layer, background])=>{
  const root = document.querySelector(layer);
  const behind = document.querySelector(background);
  const active = document.activeElement;
  return {
    open: Boolean(root && root.hasAttribute('data-open')),
    inside: Boolean(root && active && root.contains(active)),
    active: active ? (active.id || active.className || active.tagName) : null,
    backgroundInert: Boolean(behind && behind.hasAttribute('inert')),
  };}"""

# The error surfaces of one state, and how many of them announce. Hosted in a
# triple-quoted string on purpose: `check-markup-contracts.py` reads this file
# as TEXT to pair every `data-*` a rule selects with the markup that emits it,
# and an escaped quote hides the selection from it — the arm would count one
# fewer and say nothing.
SURFACES = """()=>{
  const all = [...document.querySelectorAll('[data-part="surface-error"]')];
  return [all.length,
    all.filter((node)=>node.getAttribute('role') === 'alert').length];}"""

# Where focus must be once a layer has closed: on the control that opened it,
# or — when closing the layer re-rendered that control out of existence, which
# is what happens to a poster behind a screen — on the main region. Never on
# `<body>`, which is the caret being dropped.
RESTORED = """(selector)=>{
  const trigger = document.querySelector(selector);
  const active = document.activeElement;
  if (trigger && active === trigger) return "the trigger";
  if (active && active.id === "port") return "the main region (trigger re-rendered)";
  return active ? (active.id || active.tagName) : "nothing";}"""


async def main():
    journal = Journal("R81 — focus enters a layer, and the interface says what it is doing")
    async with async_playwright() as playwright:
        browser = await playwright.chromium.launch(channel="chrome")
        context, page = await open_page(browser)

        for name, opener, layer in LAYERS:
            await page.click(opener)
            await page.wait_for_timeout(400)
            opened = await page.evaluate(FOCUS_STATE, [layer, BACKGROUND])
            journal.check(
                f"opening the {name} moves focus into it",
                opened["open"] and opened["inside"], str(opened))
            journal.check(
                f"opening the {name} takes the background out of the tab order",
                opened["backgroundInert"],
                f"{BACKGROUND} inert: {opened['backgroundInert']}")

            await page.keyboard.press("Escape")
            await page.wait_for_timeout(400)
            closed = await page.evaluate(FOCUS_STATE, [layer, BACKGROUND])
            journal.check(
                f"Escape closes the {name}",
                not closed["open"], str(closed))
            journal.check(
                f"closing the {name} gives the background back",
                not closed["backgroundInert"],
                f"{BACKGROUND} inert: {closed['backgroundInert']}")
            landed_on = await page.evaluate(RESTORED, opener)
            journal.check(
                f"closing the {name} puts focus back somewhere the reader was",
                landed_on.startswith("the "),
                f"focus went to {landed_on}")

        # ── EVERY CONFIRMATION OPENS ON ITS WAY OUT, NEVER ON THE ACT ──────
        #
        # `app/focus.ts` moves focus into a layer that opens and takes the
        # first control the reader would reach. On a confirmation that first
        # control is the ACT, and a dialog exists precisely because the act is
        # one nobody should take by accident: the restart confirmation cuts the
        # service for the whole household, and opening on « Redémarrer » hands
        # a keyboard's or a switch control's next Enter the very thing the
        # dialog was added to prevent.
        #
        # READ AS THE FOCUS, NOT AS THE MARKUP. An `autofocus` attribute in the
        # tree proves the intention; what this holds is where focus actually
        # went, because the attribute is only half a contract and the other
        # half is a layer's entry selector that may stop reading it.
        # ALL THREE THE PROTOTYPE HAS, and it used to be one. Two readers of one
        # review disagreed about the library's bulk confirmation — one had it
        # focusing « Supprimer et arrêter le suivi », the other « Annuler » —
        # and neither could be settled by a rule, because the rule walked the
        # settings dialog alone. A property held on one instance of a layer is a
        # property nobody holds.
        #
        # THE RESTART ONE IS RAISED BY A FINGER, the other two by their named
        # states: they are reached from a swipe and from selection mode, which
        # are gestures other rules own, and a walk that re-drives them here
        # would measure those gestures rather than this property.
        DIALOGS = (
            ("the restart confirmation", "settings-restart", "[data-restart]"),
            ("the library's row delete", "lib-delete", None),
            ("the library's bulk delete", "lib-delete-multiple", None),
        )
        for what, state, opener in DIALOGS:
            await page.evaluate("(id)=>window.__go(id)", state)
            await page.wait_for_timeout(500)
            if opener is not None:
                await page.evaluate(
                    """(sel)=>{const act = document.querySelector(sel);
                       if (act) act.click();}""", opener)
                await page.wait_for_timeout(500)
            landed_in_dialog = await page.evaluate("""()=>{
              const dialog = document.getElementById('dlg');
              const active = document.activeElement;
              const buttons = [...dialog.querySelectorAll('[data-part="dialog/button"]')];
              return {open: dialog.hasAttribute('data-open'),
                      inside: !!active && dialog.contains(active),
                      text: (active?.textContent || '').trim(),
                      tone: active?.dataset?.tone || 'neutral',
                      isTheWayOut: !!active && 'dialogDismiss' in active.dataset,
                      dangerous: buttons.filter(
                        (b) => b.dataset.tone === 'danger').length,
                      offered: buttons.map((b) => (b.textContent || '').trim())};}""")
            journal.check(
                f"{what} is raised, and offers a way out",
                landed_in_dialog["open"] and len(landed_in_dialog["offered"]) >= 2,
                str(landed_in_dialog["offered"]))
            journal.check(
                f"{what} really offers a destructive act, so this is not a "
                f"vacuous walk",
                landed_in_dialog["dangerous"] >= 1,
                f"{landed_in_dialog['dangerous']} danger button(s)")
            journal.check(
                f"{what} opens with focus on its WAY OUT, not on the "
                f"destructive act (B-300, §17)",
                landed_in_dialog["inside"] and landed_in_dialog["isTheWayOut"]
                and landed_in_dialog["tone"] != "danger",
                f"focus on {landed_in_dialog['text']!r}, tone "
                f"{landed_in_dialog['tone']}, way out: "
                f"{landed_in_dialog['isTheWayOut']}")
            await page.keyboard.press("Escape")
            await page.wait_for_timeout(350)

        # THE SKIP LINK, on a FRESH PAGE, which is also how a reader meets it.
        #
        # Not on the page above, and the reason is a real property of browsers
        # rather than tidiness: the sequential focus navigation starting point
        # is set by the last CLICK, and `blur()` does not move it. After the
        # layer holds — which deliberately leave focus on the control that
        # opened them — Tab resumes after the avatar, so the first version of
        # this hold measured the fourth stop and reported the skip link
        # missing. A fresh page has no starting point but the document's.
        fresh_context, fresh = await open_page(browser)
        await fresh.keyboard.press("Tab")
        await fresh.wait_for_timeout(150)
        first = await fresh.evaluate(
            "()=>document.activeElement?.className || ''")
        journal.check(
            "the first stop of the tab order is the skip link",
            "skip-link" in first, f"active: {first!r}")
        await fresh.keyboard.press("Enter")
        await fresh.wait_for_timeout(250)
        landed = await fresh.evaluate(
            "()=>document.activeElement?.id || ''")
        journal.check(
            "following the skip link moves FOCUS to the main region, not only "
            "the scroll position",
            landed == "port", f"active: {landed!r}")
        await fresh_context.close()

        # WHAT THE INTERFACE SAYS WHILE IT WORKS, here for the same reason as
        # the rest: an audit reads a moment, and « busy » is a moment it is not
        # driving.
        #
        # The busy mark is set in ONE place — the page host, the only thing that
        # knows every page's phase. Marked page by page it would be eight call
        # sites, and the eighth would be forgotten.
        await page.evaluate("(id)=>window.__go(id)", "lib-loading")
        await page.wait_for_timeout(400)
        journal.check(
            "a page that is loading says so",
            await page.evaluate(
                "()=>document.querySelector('#port')?.getAttribute('aria-busy')"
            ) == "true",
            "aria-busy while loading")
        await page.evaluate("(id)=>window.__go(id)", "lib-grid")
        await page.wait_for_timeout(400)
        journal.check(
            "and stops saying so once it has loaded",
            await page.evaluate(
                "()=>document.querySelector('#port')?.hasAttribute('aria-busy')"
            ) is False,
            "aria-busy cleared")

        # EVERY error surface announces, over EVERY state that renders one.
        #
        # The shape is repeated at nine call sites across six files, which is
        # exactly where the ninth gets forgotten — so this drives every named
        # state whose id says error and sums what it finds. A first version of
        # this hold drove one state, found one surface and called that a count:
        # a hold that samples and reads like a census is worse than no hold,
        # because it is believed.
        seen, announced = 0, 0
        for state in [s for s in await page.evaluate("()=>window.__states()")
                      if "error" in s]:
            await page.evaluate("(id)=>window.__go(id)", state)
            await page.wait_for_timeout(300)
            found = await page.evaluate(SURFACES)
            seen += found[0]
            announced += found[1]
        journal.check(
            "every error surface reaches a listener, not only a reader",
            seen > 0 and seen == announced,
            f"{announced} of {seen} surface(s) announce, over the error states")

        await context.close()
        await browser.close()
    print(BAR)
    journal.summary()


if __name__ == "__main__":
    sys.exit(asyncio.run(main()) or 0)

File: frontend/maquette/design/src/app/focus.ts
// design/src/focus.ts
// FOCUS, WHEN A LAYER OPENS AND WHEN IT CLOSES.
//
// A layer that opens without taking focus leaves a keyboard on the page behind
// it: Tab walks a list the reader cannot see, Escape has nothing to close, and
// a screen reader keeps announcing a surface the interface has covered. A layer
// that closes without GIVING focus back is worse — the caret lands at the top
// of the document and the reader begins again.
//
// WHAT IT OBSERVES, AND WHY IT NEEDS NOTHING FROM THE ENGINE. Every layer in
// this prototype already announces itself: `setOpen(element, on)` toggles the
// `data-open` attribute on `#drawer`, `#screen`, `#dlg` and `#scrim`, and the
// React side (`components/sheet.tsx`, the five screens) emits the same
// attribute itself. So this module watches an attribute that exists rather than
// asking anyone to call it. Nothing in the engine changes; the contract is
// markup, which is what makes this survive the engine's removal.
//
// WHY `inert` AND NOT `aria-hidden`. `aria-hidden` hides a subtree from a
// screen reader and leaves every control in it TABBABLE — the reader tabs into
// something that is no longer announced, which is a worse state than either
// half alone. `inert` removes the subtree from the tab order AND from the
// accessibility tree, in one attribute the platform implements.
//
// THE BACKGROUND IS THE FRAME'S OTHER CHILDREN, never the layer's ancestors.
// Marking `document.body` inert would mark the layer too.

// The layer roots, in the stacking order the engine already unwinds — drawer,
// then screen, then sheet — so the topmost open layer is the last one here that
// carries `data-open`.
//
// A SCREEN IS SELECTED BY ITS PART, NEVER BY `#screen`. There is one legacy
// screen with that id and FIVE migrated ones that are `<section data-part=
// "screen">` with no id at all, so an id selector traps focus in the layer the
// engine still draws and silently ignores every screen that has been converted
// — which is most of them, and all of the ones a later lot will add. This was
// the shape of the first version of this file.
const LAYERS = [
  "#drawer",
  '[data-part="screen"]',
  "#sheet",
  "#dlg",
] as const;

// A layer's own named entry point, asked for FIRST and on its own.
//
// IT USED TO BE THE FIRST ALTERNATIVE OF `ENTRY`, under a comment saying
// « `[autofocus]` first, so a layer can name its own entry point » — and a
// selector LIST has no priority: `querySelector` answers the first matching
// node in DOCUMENT order, whatever order the alternatives are written in. So a
// dialog whose destructive button comes before its way out got focus on the
// destructive button while carrying `autofocus` on the other, and the comment
// asserting otherwise had been there the whole time. Measured: the attribute
// rendered, `[autofocus]` matched one node, and `document.activeElement` was
// the other button.
const NAMED_ENTRY = "[autofocus]";

// Where focus goes when a layer names none: the first control the reader would
// reach anyway.
const ENTRY =
  "button:not([disabled]),[href],input:not([disabled])," +
  "select:not([disabled]),textarea:not([disabled]),[tabindex]:not([tabindex='-1'])";

type Restore = { layer: Element; trigger: HTMLElement | null };

let stack: Restore[] = [];
let observer: MutationObserver | null = null;

function isOpen(node: Element): boolean {
  // `isConnected` FIRST, and it is not defensive noise. The drawer, the sheet
  // and the dialog stay in the document and lose `data-open` when they close;
  // a migrated screen is UNMOUNTED instead, and a detached node keeps every
  // attribute it had. Reading the attribute alone, the stack held a layer that
  // was open forever, nothing was ever unwound, and focus stayed where the
  // vanished screen had left it — on `<body>`.
  return node.isConnected && node.hasAttribute("data-open");
}

function openLayers(): Element[] {
  // `querySelectorAll`, because a selector here can match several elements:
  // the legacy screen and every migrated one answer to the same part.
  return LAYERS.flatMap((selector) =>
    [...document.querySelectorAll(selector)].filter(isOpen),
  );
}

/**
 * Marks everything except the open layer as `inert`, or clears it.
 *
 * The siblings walked are the phone frame's children — the shell's header, the
 * main region, the tab bar and the other layers. An ancestor can never be
 * inert: it contains the layer.
 *
 * @param layer The layer that owns focus, or null to clear every mark.
 */
function setBackgroundInert(layer: Element | null): void {
  const frame = document.getElementById("device");
  if (!frame) return;
  for (const child of Array.from(frame.children)) {
    // The React mount node holds the sheet AND the screens; marking it whole
    // would mark the very layer that is open. Its children are walked instead.
    const nodes = child.id === "shell" ? Array.from(child.children) : [child];
    for (const node of nodes) {
      // THE SCRIM IS NEVER INERT. It is not background — it is the layer's own
      // backdrop, and tapping it is how the layer closes on a phone. Marked
      // inert it stops receiving pointer events at all, so a tap outside the
      // drawer did nothing: `drawer.py` caught it as history that would not
      // move, which is what that hold is for.
      if (node.id === "scrim") {
        node.removeAttribute("inert");
        continue;
      }
      const contains = layer ? node === layer || node.contains(layer) : false;
      if (layer && !contains) node.setAttribute("inert", "");
      else node.removeAttribute("inert");
    }
  }
}

/**
 * Moves focus into a layer that has just opened.
 *
 * @param layer The layer's root.
 */
function focusInto(layer: Element): void {
  const target = layer.querySelector<HTMLElement>(NAMED_ENTRY)
    ?? layer.querySelector<HTMLElement>(ENTRY);
  if (target) {
    target.focus();
    return;
  }
  // A layer with no control of its own still has to receive focus, or the
  // reader stays behind it. `tabindex="-1"` makes the root focusable by script
  // without adding a stop to the tab order.
  const root = layer as HTMLElement;
  root.setAttribute("tabindex", "-1");
  root.focus();
}

/** Reconciles the focus stack with what the markup currently says is open. */
function reconcile(): void {
  const open = openLayers();
  const top = open.length ? open[open.length - 1] : null;
  const previous = stack.length ? stack[stack.length - 1] : null;

  if (top && previous?.layer === top) return;

  // Closing: unwind every entry whose layer is no longer open, giving focus
  // back to the trigger recorded when it opened.
  //
  // THE INERT MARK COMES OFF FIRST, and that ordering is the whole of it.
  // `focus()` on an element inside an `inert` subtree does nothing — silently,
  // with no error and no return value to check — and every trigger this
  // restores is in the background the layer had just made inert. Restoring
  // before clearing left the caret on `<body>` at every close, which is exactly
  // the defect the manager exists to prevent, reproduced by the manager.
  const unwinding = stack.length && !isOpen(stack[stack.length - 1].layer);
  if (unwinding && !top) setBackgroundInert(null);
  while (stack.length && !isOpen(stack[stack.length - 1].layer)) {
    const closed = stack.pop() as Restore;
    if (open.length) continue;
    // THE TRIGGER MAY NOT EXIST ANY MORE, and for a screen it usually does not:
    // it was a poster in the library, and closing the screen re-renders the
    // list, so the node that was clicked is replaced by an equal-looking new
    // one. Focusing a detached element does nothing, so the caret would stay on
    // `<body>` — the very outcome this whole module exists to prevent, arrived
    // at by a different road than the `inert` one.
    //
    // The fallback is the main region rather than nothing: the reader lands
    // back in the content they came from instead of at the top of the document.
    if (closed.trigger?.isConnected) closed.trigger.focus();
    else document.getElementById("port")?.focus();
  }

  if (!top) {
    setBackgroundInert(null);
    return;
  }
  if (!stack.some((entry) => entry.layer === top)) {
    const active = document.activeElement;
    stack.push({
      layer: top,
      // The trigger is recorded BEFORE focus moves. `document.body` is what
      // `activeElement` reports when nothing is focused, and giving focus back
      // to the body is the same as giving it back to nowhere.
      trigger:
        active instanceof HTMLElement && active !== document.body ? active : null,
    });
    setBackgroundInert(top);
    focusInto(top);
    return;
  }
  setBackgroundInert(top);
}

/**
 * Installs the layer focus manager. Idempotent — a second call is a no-op.
 *
 * Called once from the shell's boot, before the engine starts, so the first
 * layer an operator opens is already covered.
 */
export function installFocusManager(): void {
  if (observer) return;
  observer = new MutationObserver(reconcile);
  observer.observe(document.body, {
    subtree: true,
    childList: true,
    attributes: true,
    attributeFilter: ["data-open"],
  });
  // THE SKIP LINK IS DRIVEN HERE, not left to the browser. `<a href="#port">`
  // should move focus to a target carrying `tabindex="-1"`, and measured, it
  // does not in this prototype: the router owns the URL and the hash never
  // reaches the document's default handling, so the link scrolled and the caret
  // stayed on `<body>` — a skip link that moves the view and not the focus is
  // the commonest way to ship one that does nothing.
  document.addEventListener("click", (event) => {
    const target = event.target as Element | null;
    const link = target?.closest?.('[data-part="shell/skip-link"]');
    if (!link) return;
    event.preventDefault();
    const main = document.getElementById("port");
    if (!main) return;
    main.focus();
    main.scrollTop = 0;
  });

  // `keydown` on the document, in the CAPTURE phase: a layer's own handler must
  // not be able to swallow the one key that closes it.
  document.addEventListener(
    "keydown",
    (event) => {
      if (event.key !== "Escape") return;
      const open = openLayers();
      if (!open.length) return;
      event.preventDefault();
      // TWO VERBS, because the layers close two different ways and pretending
      // otherwise leaves one of them stuck. `__closeLayers` closes the dialog,
      // the sheet and the drawer — the three the scrim covers — and it
      // deliberately does not close a SCREEN: a screen is a history entry, and
      // what closes one is a back. Sending Escape to `__closeLayers` on a
      // screen did nothing at all, silently, with the background still inert.
      //
      // Both verbs are the ones the interface already has. Nothing new is
      // written here: a second closer would give one gesture two answers.
      const top = open[open.length - 1];
      if (top.matches('[data-part="screen"]')) window.__bridge?.back();
      else window.__closeLayers?.();
    },
    true,
  );
  reconcile();
}
